source/adapters/level_zero/queue.cpp

```diff
@@ -1414,7 +1435,16 @@ ur_result_t ur_queue_handle_t_::synchronize() {
     // zero handle can have device scope, so we can't synchronize the last
     // event.
     if (isInOrderQueue() && !LastCommandEvent->IsDiscarded) {
-      ZE2UR_CALL(zeHostSynchronize, (LastCommandEvent->ZeEvent));
+      if (UrL0QueueSyncNonBlocking) {
+        this->Mutex.unlock();
+        {
+          std::scoped_lock<ur_shared_mutex> Lock(this->zeQueueInUseMutex);
```

Comment:
dropping the `Mutex` and acquiring the `zeQueueInUseMutex` would have to be atomic for this to be safe. Otherwise, assuming RefCount can be 0 here, a thread calling `urQueueReleaseInternal` could delete the queue object in the unsynchronized gap between these locks.

But, again, I don't believe `zeQueueInUseMutex` is even necessary.

Reply:
I had added the mutex because originally I was not sure, but let me check again if there is any possibility. If there isn't, then I will remove the mutex all together and instead have the "queue sync" be in a "lockless" mode while awaiting the return from the sync.